File: common/src/main/java/com/matejdro/wearremotelist/receiverside/conn/PlayServicesConnectionToProvider.java
package com.matejdro.wearremotelist.receiverside.conn;

import android.net.Uri;
import android.os.AsyncTask;
import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.Nullable;

import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.wearable.MessageApi;
import com.google.android.gms.wearable.MessageEvent;
import com.google.android.gms.wearable.Wearable;
import com.matejdro.wearremotelist.CommonPlayServicesPaths;
import com.matejdro.wearremotelist.ErrorListener;
import com.matejdro.wearremotelist.receiverside.RemoteListManager;

import java.nio.ByteBuffer;

/**
 * Class that connects to the provider via Google Play Services's MessageApi.
 */
public abstract class PlayServicesConnectionToProvider implements ConnectionToProvider, MessageApi.MessageListener
{
    private GoogleApiClient googleApiClient;
    private RemoteListManager.ListDataReceiver dataReceiver;

    /**
     * @param googleApiClient Connected GoogleApiClient with Wearable API enabled.
     */
    public PlayServicesConnectionToProvider(GoogleApiClient googleApiClient)
    {
        if (googleApiClient == null)
            throw new NullPointerException();

        if (!googleApiClient.isConnected())
            throw new IllegalArgumentException("Provided GoogleApiClient is not connected to Google Play Services.");
        if (!googleApiClient.hasConnectedApi(Wearable.API))
            throw new IllegalArgumentException("Provided GoogleApiClient does not have Wearable API connected.");

        this.googleApiClient = googleApiClient;

        Wearable.MessageApi.addListener(googleApiClient, this);
    }

    /**
     * Remove my listener from provided GoogleApiClient.
     */
    public void disconnect()
    {
        Wearable.MessageApi.removeListener(googleApiClient, this);
    }

    @Override
    public void setDataReceiver(RemoteListManager.ListDataReceiver dataReceiver)
    {
        this.dataReceiver = dataReceiver;
    }

    public GoogleApiClient getGoogleApiClient()
    {
        return googleApiClient;
    }

    public abstract @Nullable String getProviderNodeId();

    @Override
    public void requestListSize(final String listPath)
    {
        if (dataReceiver == null)
            throw new IllegalStateException("Data Receiver is still null when transferring messages");

        new AsyncTask<Void, Void, Void>()
        {
            @Override
            protected Void doInBackground(Void... voids)
            {
                String providerNodeId = getProviderNodeId();
                if (providerNodeId == null)
                {
                    dataReceiver.onError(listPath, ErrorListener.ERROR_DISCONNECTED);
                    return null;
                }

                Uri targetUri = CommonPlayServicesPaths.REQUEST_LIST_SIZE_URI
                        .buildUpon()
                        .appendQueryParameter(CommonPlayServicesPaths.PARAMETER_LIST_PATH, listPath)
                        .build();

                Wearable.MessageApi.sendMessage(googleApiClient, providerNodeId, targetUri.toString(), null)
                        .setResultCallback(new ListMessageResultCallback(listPath));

                return null;
            }
        }.execute((Void[]) null);
    }

    @Override
    public void requestItems(final String listPath, final int from, final int to)
    {
        if (dataReceiver == null)
            throw new IllegalStateException("Data Receiver is still null when transferring messages");

        new AsyncTask<Void, Void, Void>()
        {
            @Override
            protected Void doInBackground(Void... voids)
            {
                String providerNodeId = getProviderNodeId();
                if (providerNodeId == null)
                {
                    dataReceiver.onError(listPath, ErrorListener.ERROR_DISCONNECTED);
                    return null;
                }

                Uri targetUri = CommonPlayServicesPaths.REQUEST_ITEMS_URI
                        .buildUpon()
                        .appendQueryParameter(CommonPlayServicesPaths.PARAMETER_LIST_PATH, listPath)
                        .appendQueryParameter(CommonPlayServicesPaths.PARAMETER_ITEMS_FROM, Integer.toString(from))
                        .appendQueryParameter(CommonPlayServicesPaths.PARAMETER_ITEMS_TO, Integer.toString(to))

                        .build();

                Wearable.MessageApi.sendMessage(googleApiClient, providerNodeId, targetUri.toString(), null)
                        .setResultCallback(new ListMessageResultCallback(listPath));

                return null;
            }
        }.execute((Void[]) null);
    }

    @Override
    public void onMessageReceived(MessageEvent messageEvent)
    {
        if (dataReceiver == null)
            throw new IllegalStateException("Data Receiver is still null when messages are arriving.");

        Uri messageUri = Uri.parse(messageEvent.getPath());
        if (CommonPlayServicesPaths.RESPONSE_LIST_SIZE_URI.getPath().equals(messageUri.getPath()))
        {

            String listPath = messageUri.getQueryParameter(CommonPlayServicesPaths.PARAMETER_LIST_PATH);
            int newSize = ByteBuffer.wrap(messageEvent.getData()).getInt();

            if (newSize < 0)
            {
                dataReceiver.onError(listPath, ErrorListener.ERROR_UNKNOWN_LIST);
            }
            else
            {
                dataReceiver.updateSizeReceived(listPath, newSize);
            }
        }
        else if (CommonPlayServicesPaths.RESPONSE_ITEMS_URI.getPath().equals(messageUri.getPath()))
        {
            String listPath = messageUri.getQueryParameter(CommonPlayServicesPaths.PARAMETER_LIST_PATH);
            int from = Integer.parseInt(messageUri.getQueryParameter(CommonPlayServicesPaths.PARAMETER_ITEMS_FROM));
            int to = Integer.parseInt(messageUri.getQueryParameter(CommonPlayServicesPaths.PARAMETER_ITEMS_TO));

            Parcel parcel = Parcel.obtain();
            parcel.unmarshall(messageEvent.getData(), 0, messageEvent.getData().length);
            parcel.setDataPosition(0);

            Parcelable.Creator creator = dataReceiver.getParcelableCreator(listPath);
            Parcelable[] itemsSent = new Parcelable[to - from + 1];
            for (int i = 0; i < itemsSent.length; i++)
            {
                itemsSent[i] = (Parcelable) creator.createFromParcel(parcel);
            }

            parcel.recycle();
            dataReceiver.dataReceived(listPath, from, itemsSent);
        }
    }

    private class ListMessageResultCallback implements ResultCallback<MessageApi.SendMessageResult>
    {
        private String listPath;

        public ListMessageResultCallback(String listPath)
        {
            this.listPath = listPath;
        }

        @Override
        public void onResult(MessageApi.SendMessageResult sendMessageResult)
        {
            if (!sendMessageResult.getStatus().isSuccess())
            {
                if (dataReceiver != null)
                    dataReceiver.onError(listPath, CommonPlayServicesPaths.getErrorCodeFromGMSStatus(sendMessageResult.getStatus().getStatusCode()));
            }
        }
    }
}

File: common/src/main/java/com/matejdro/wearremotelist/receiverside/conn/PlayServicesConnectionToProvider.kt
package com.matejdro.wearremotelist.receiverside.conn

import android.net.Uri
import android.os.AsyncTask
import android.os.Parcel
import android.os.Parcelable
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.common.api.ResultCallback
import com.google.android.gms.wearable.MessageApi.MessageListener
import com.google.android.gms.wearable.MessageApi.SendMessageResult
import com.google.android.gms.wearable.MessageEvent
import com.google.android.gms.wearable.Wearable
import com.matejdro.wearremotelist.CommonPlayServicesPaths
import com.matejdro.wearremotelist.ErrorListener
import com.matejdro.wearremotelist.receiverside.RemoteListManager.ListDataReceiver
import java.nio.ByteBuffer

/**
 * Class that connects to the provider via Google Play Services's MessageApi.
 */
abstract class PlayServicesConnectionToProvider(googleApiClient: GoogleApiClient?) :
    ConnectionToProvider, MessageListener {
    val googleApiClient: GoogleApiClient
    private var dataReceiver: ListDataReceiver? = null

    /**
     * @param googleApiClient Connected GoogleApiClient with Wearable API enabled.
     */
    init {
        if (googleApiClient == null) throw NullPointerException()
        require(googleApiClient.isConnected) { "Provided GoogleApiClient is not connected to Google Play Services." }
        require(googleApiClient.hasConnectedApi(Wearable.API)) { "Provided GoogleApiClient does not have Wearable API connected." }
        this.googleApiClient = googleApiClient
        Wearable.MessageApi.addListener(googleApiClient, this)
    }

    /**
     * Remove my listener from provided GoogleApiClient.
     */
    fun disconnect() {
        Wearable.MessageApi.removeListener(googleApiClient, this)
    }

    override fun setDataReceiver(dataReceiver: ListDataReceiver) {
        this.dataReceiver = dataReceiver
    }

    abstract fun getProviderNodeId(): String?
    override fun requestListSize(listPath: String) {
        checkNotNull(dataReceiver) { "Data Receiver is still null when transferring messages" }
        object : AsyncTask<Void?, Void?, Void?>() {
            @Deprecated("Deprecated in Java")
            protected override fun doInBackground(vararg params: Void?): Void? {
                val providerNodeId: String = getProviderNodeId()!!
                if (providerNodeId == null) {
                    dataReceiver!!.onError(listPath, ErrorListener.ERROR_DISCONNECTED)
                    return null
                }
                val targetUri = CommonPlayServicesPaths.REQUEST_LIST_SIZE_URI
                    .buildUpon()
                    .appendQueryParameter(CommonPlayServicesPaths.PARAMETER_LIST_PATH, listPath)
                    .build()
                Wearable.MessageApi.sendMessage(
                    googleApiClient,
                    providerNodeId,
                    targetUri.toString(),
                    byteArrayOf()
                )
                    .setResultCallback(ListMessageResultCallback(listPath))
                return null
            }
        }.execute(null)
    }

    override fun requestItems(listPath: String, from: Int, to: Int) {
        checkNotNull(dataReceiver) { "Data Receiver is still null when transferring messages" }
        object : AsyncTask<Void?, Void?, Void?>() {
            @Deprecated("Deprecated in Java")
            protected override fun doInBackground(vararg params: Void?): Void? {
                val providerNodeId: String = getProviderNodeId()!!
                if (providerNodeId == null) {
                    dataReceiver!!.onError(listPath, ErrorListener.ERROR_DISCONNECTED)
                    return null
                }
                val targetUri = CommonPlayServicesPaths.REQUEST_ITEMS_URI
                    .buildUpon()
                    .appendQueryParameter(CommonPlayServicesPaths.PARAMETER_LIST_PATH, listPath)
                    .appendQueryParameter(
                        CommonPlayServicesPaths.PARAMETER_ITEMS_FROM,
                        from.toString()
                    )
                    .appendQueryParameter(CommonPlayServicesPaths.PARAMETER_ITEMS_TO, to.toString())
                    .build()
                Wearable.MessageApi.sendMessage(
                    googleApiClient,
                    providerNodeId,
                    targetUri.toString(),
                    byteArrayOf()
                )
                    .setResultCallback(ListMessageResultCallback(listPath))
                return null
            }
        }.execute(null)
    }

    override fun onMessageReceived(messageEvent: MessageEvent) {
        checkNotNull(dataReceiver) { "Data Receiver is still null when messages are arriving." }
        val messageUri = Uri.parse(messageEvent.path)
        if (CommonPlayServicesPaths.RESPONSE_LIST_SIZE_URI.path == messageUri.path) {
            val listPath = messageUri.getQueryParameter(CommonPlayServicesPaths.PARAMETER_LIST_PATH)
            val newSize = ByteBuffer.wrap(messageEvent.data).getInt()
            if (newSize < 0) {
                dataReceiver!!.onError(listPath, ErrorListener.ERROR_UNKNOWN_LIST)
            } else {
                dataReceiver!!.updateSizeReceived(listPath, newSize)
            }
        } else if (CommonPlayServicesPaths.RESPONSE_ITEMS_URI.path == messageUri.path) {
            val listPath = messageUri.getQueryParameter(CommonPlayServicesPaths.PARAMETER_LIST_PATH)
            val from = messageUri.getQueryParameter(CommonPlayServicesPaths.PARAMETER_ITEMS_FROM)!!
                .toInt()
            val to = messageUri.getQueryParameter(CommonPlayServicesPaths.PARAMETER_ITEMS_TO)!!
                .toInt()
            val parcel = Parcel.obtain()
            parcel.unmarshall(messageEvent.data, 0, messageEvent.data.size)
            parcel.setDataPosition(0)
            val creator = dataReceiver!!.getParcelableCreator(listPath)
            val itemsSent = arrayOfNulls<Parcelable>(to - from + 1)
            for (i in itemsSent.indices) {
                itemsSent[i] = creator.createFromParcel(parcel) as Parcelable
            }
            parcel.recycle()
            dataReceiver!!.dataReceived(listPath, from, itemsSent)
        }
    }

    private inner class ListMessageResultCallback(private val listPath: String) :
        ResultCallback<SendMessageResult> {
        override fun onResult(sendMessageResult: SendMessageResult) {
            if (!sendMessageResult.status.isSuccess) {
                if (dataReceiver != null) dataReceiver!!.onError(
                    listPath,
                    CommonPlayServicesPaths.getErrorCodeFromGMSStatus(sendMessageResult.status.statusCode)
                )
            }
        }
    }
}
